pkg/controllers/sync/version/manager.go

```diff
@@ -23,44 +22,38 @@ package version
 
 import (
 	"context"
-	"fmt"
-	"strings"
 	"sync"
 	"time"
 
 	"github.com/pkg/errors"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/apimachinery/pkg/api/meta"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
-	pkgruntime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
 	"k8s.io/apimachinery/pkg/util/sets"
 	"k8s.io/apimachinery/pkg/util/wait"
 	"k8s.io/klog/v2"
 	runtimeclient "sigs.k8s.io/controller-runtime/pkg/client"
 
 	fedcorev1a1 "github.com/kubewharf/kubeadmiral/pkg/apis/core/v1alpha1"
-	"github.com/kubewharf/kubeadmiral/pkg/client/generic"
+	fedcorev1a1client "github.com/kubewharf/kubeadmiral/pkg/client/clientset/versioned/typed/core/v1alpha1"
 	"github.com/kubewharf/kubeadmiral/pkg/controllers/common"
-	"github.com/kubewharf/kubeadmiral/pkg/controllers/util"
+	"github.com/kubewharf/kubeadmiral/pkg/util/propagatedversion"
 )
 
 // VersionedResource defines the methods a federated resource must
 // implement to allow versions to be tracked by the VersionManager.
 type VersionedResource interface {
 	FederatedName() common.QualifiedName
-	Object() *unstructured.Unstructured
+	Object() fedcorev1a1.GenericFederatedObject
 	TemplateVersion() (string, error)
 	OverrideVersion() (string, error)
+	FederatedGVK() schema.GroupVersionKind
 }
 
 type VersionManager struct {
```

Comment:
Let's add some comments to explain what VersionManager does.

> VersionManager is used by the Sync controller to record the last synced version of a FederatedObject along with the versions of the cluster objects that were created or updated in the process. This is used to prevent sending unnecessary update requests to member clusters in subsequent reconciliations. The information is persisted in the apiserver as PropagatedVersion and ClusterPropagated CRs.
>
> In the context of the Sync controller, we identify the "version" of a FederatedObject with the hash of its template and overrides and we identify the "version" of a cluster object to be either its Generation (if available) or its ResourceVersion.
> 
> VersionManager is required because created/updated cluster objects might not match the template directly due to default values, admission plugins or webhooks. Thus we have to store the version returned by the create/update request to avoid false-positives when determining if the cluster object has diverged from the template.